pkg/expression/builtin_time.go

```diff
@@ -67,7 +67,34 @@ var (
 	durationPattern = regexp.MustCompile(`^\s*[-]?(((\d{1,2}\s+)?0*\d{0,3}(:0*\d{1,2}){0,2})|(\d{1,7}))?(\.\d*)?\s*$`)
 
 	// timestampPattern checks whether a string matches the format of timestamp.
-	timestampPattern = regexp.MustCompile(`^\s*0*\d{1,4}([^\d]0*\d{1,2}){2}\s+(0*\d{0,2}([^\d]0*\d{1,2}){2})?(\.\d*)?\s*$`)
+	timestampPattern = regexp.MustCompile(`^` +
+		// Skip any spaces or zeros
+		`\s*0*` +
+		// Year 1-4 digits
+		`\d{1,4}` +
+		// TODO: Add warning if non '-' separator in ParseTime
+		// 1 or 2 digit Month and Day
+		// Any non-digit as separator
+		// Any leading 0's for Month/Day
+		`([^\d]0*\d{1,2}){2}` +
+		// At least one space between Date and Time parts
+		`\s+` +
+		// Hour is mandatory
+		// Any number of leading zeroes
+		// 1-2 Hour digits
+		`0*\d{1,2}` +
+		// Minutes or Minutes:Seconds are optional
+		// Any non-digit separator before Minute and Second parts
+		// Any number of leading zeroes in Min/Sec!
+		// 1-2 digit minutes/seconds
+		`([^\d]0*\d{1,2}){0,2}` +
+		// Optionally decimal comma (.) and 0 or more fractional seconds
+		// (regardless if min/sec exists or not...)
+		`(\.\d*)?` +
+		// Optionally time zone offset, must be +/-HH:MM format
+		`([+-]\d{2}[:]\d{2})?` +
```

Comment:
MySQL has this part very strict. It is more strict than what types.ParseTime parses the same string, so I think we should be more strict here as well, to avoid the situation where customer starts to use a more relaxed syntax, and we need to make it more strict.

Reply:
Yes, I agree with that